### PR TITLE
[pulsar]: support upgrade to 2.10 image base to pulsar chart

### PR DESCRIPTION
**Pull request by ericsyh**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

* upgrade the pulsar image to `2.10.3.6`
* add the required `securityContext` configuration for zk and bk
* add some probe configs to zk and broker

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: charts/pulsar/templates/broker/broker-statefulset.yaml [hunk 1]
```diff
@@ -219,6 +219,7 @@ spec:
           initialDelaySeconds: {{ .Values.broker.probe.liveness.initialDelaySeconds }}
           periodSeconds: {{ .Values.broker.probe.liveness.periodSeconds }}
           failureThreshold: {{ .Values.broker.probe.liveness.failureThreshold }}
+          timeoutSeconds: {{ .Values.broker.probe.liveness.timeoutSeconds }}
```
Comment (heesung-sn) [inline charts/pulsar/templates/broker/broker-statefulset.yaml:222]: why do we add this timeoutSeconds?

File: charts/pulsar/templates/zookeeper/zookeeper-statefulset.yaml [hunk 2]
```diff
@@ -41,7 +41,7 @@ spec:
       component: {{ .Values.zookeeper.component }}
   updateStrategy:
     type: {{ .Values.zookeeper.updateStrategy }}
-  podManagementPolicy: {{ .Values.zookeeper.podManagementPolicy | default "OrderedReady" }}
+  podManagementPolicy: {{ .Values.zookeeper.podManagementPolicy | default "Parallel" }}
```
Comment (heesung-sn) [inline charts/pulsar/templates/zookeeper/zookeeper-statefulset.yaml:44]: why do we need to change the default from OrderedReady to Parallel?

File: charts/pulsar/values.yaml [hunk 3]
```diff
@@ -530,18 +530,19 @@ zookeeper:
       failureThreshold: 10
       initialDelaySeconds: 10
       periodSeconds: 30
-      timeoutSeconds: 5
+      timeoutSeconds: 10
```
Comment (heesung-sn) [inline charts/pulsar/values.yaml:533]: Why do we increase timeout from 5 to 10?

File: charts/pulsar/values.yaml [hunk 4]
```diff
@@ -551,7 +552,9 @@ zookeeper:
   annotations:
     prometheus.io/scrape: "true"
     prometheus.io/port: "8000"
-  securityContext: {}
+  securityContext: 
+    fsGroup: 0
```
Comment (heesung-sn) [inline charts/pulsar/values.yaml:556]: why is fsGroup 0? (does it have to be 0 or is it randomly selected?)

File: charts/pulsar/templates/zookeeper/zookeeper-statefulset.yaml (hunk 2 above)
Comment (ericsyh) [reply charts/pulsar/templates/zookeeper/zookeeper-statefulset.yaml:44]: I remember that's a suggstion from Yang Yang, if using pulsar operators this is not required for zk. But native statefulset ZK seems suggest to use Parallel, also refer the https://github.com/apache/pulsar-helm-chart/blob/master/charts/pulsar/values.yaml#L287

File: charts/pulsar/templates/broker/broker-statefulset.yaml (hunk 1 above)
Comment (ericsyh) [reply charts/pulsar/templates/broker/broker-statefulset.yaml:222]: In my test, I found broker will fail to start and loop in restarting for several times. I captured some logs, looks like relate to the function worker in the broker?

```
10:51:04.244 [AsyncHttpClient-53-1] WARN  org.apache.pulsar.client.admin.internal.BaseResource - [http://pulsar-broker-1.pulsar-broker.pulsar.svc.cluster.local:8080/admin/v2/persistent/public/functions/assignments] Failed to perform http put request: java.util.concurrent.CompletionException: org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector$RetryException: Could not complete the operation. Number of retries has been exhausted. Failed reason: pulsar-broker-2.pulsar-broker.pulsar.svc.cluster.local
10:51:04.245 [main] ERROR org.apache.pulsar.functions.worker.PulsarWorkerService - Error Starting up in worker
org.apache.pulsar.client.admin.PulsarAdminException: java.util.concurrent.CompletionException: org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector$RetryException: Could not complete the operation. Number of retries has been exhausted. Failed reason: pulsar-broker-2.pulsar-broker.pulsar.svc.cluster.local
	at org.apache.pulsar.client.admin.PulsarAdminException.wrap(PulsarAdminException.java:252) ~[io.streamnative-pulsar-client-admin-api-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.client.admin.internal.BaseResource.sync(BaseResource.java:302) ~[io.streamnative-pulsar-client-admin-original-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.client.admin.internal.TopicsImpl.createNonPartitionedTopic(TopicsImpl.java:340) ~[io.streamnative-pulsar-client-admin-original-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.client.admin.Topics.createNonPartitionedTopic(Topics.java:482) ~[io.streamnative-pulsar-client-admin-api-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.functions.worker.PulsarWorkerService.tryCreateNonPartitionedTopic(PulsarWorkerService.java:390) ~[io.streamnative-pulsar-functions-worker-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.functions.worker.PulsarWorkerService.start(PulsarWorkerService.java:441) ~[io.streamnative-pulsar-functions-worker-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.broker.PulsarService.startWorkerService(PulsarService.java:1624) ~[io.streamnative-pulsar-broker-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:808) ~[io.streamnative-pulsar-broker-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.PulsarBrokerStarter$BrokerStarter.start(PulsarBrokerStarter.java:273) ~[io.streamnative-pulsar-broker-2.10.3.6.jar:2.10.3.6]
	at org.apache.pulsar.PulsarBrokerStarter.main(PulsarBrokerStarter.java:350) ~[io.streamnative-pulsar-broker-2.10.3.6.jar:2.10.3.6]
	Suppressed: org.apache.pulsar.client.admin.PulsarAdminException: java.util.concurrent.CompletionException: org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector$RetryException: Could not complete the operation. Number of retries has been exhausted. Failed reason: pulsar-broker-2.pulsar-broker.pulsar.svc.cluster.local
		at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:251) ~[io.streamnative-pulsar-client-admin-original-2.10.3.6.jar:2.10.3.6]
		at org.apache.pulsar.client.admin.internal.BaseResource$1.failed(BaseResource.java:134) ~[io.streamnative-pulsar-client-admin-original-2.10.3.6.jar:2.10.3.6]
		at org.glassfish.jersey.client.JerseyInvocation$1.failed(JerseyInvocation.java:882) ~[org.glassfish.jersey.core-jersey-client-2.34.jar:?]
		at org.glassfish.jersey.client.ClientRuntime.processFailure(ClientRuntime.java:247) ~[org.glassfish.jersey.core-jersey-client-2.34.jar:?]
		at org.glassfish.jersey.client.ClientRuntime.processFailure(ClientRuntime.java:242) ~[org.glassfish.jersey.core-jersey-client-2.34.jar:?]
		at org.glassfish.jersey.client.ClientRuntime.access$100(ClientRuntime.java:62) ~[org.glassfish.jersey.core-jersey-client-2.34.jar:?]
		at org.glassfish.jersey.client.ClientRuntime$2.lambda$failure$1(ClientRuntime.java:178) ~[org.glassfish.jersey.core-jersey-client-2.34.jar:?]
		at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) ~[org.glassfish.jersey.core-jersey-common-2.34.jar:?]
		at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) ~[org.glassfish.jersey.core-jersey-common-2.34.jar:?]
		at org.glassfish.jersey.internal.Errors.process(Errors.java:292) ~[org.glassfish.jersey.core-jersey-common-2.34.jar:?]
		at org.glassfish.jersey.internal.Errors.process(Errors.java:274) ~[org.glassfish.jersey.core-jersey-common-2.34.jar:?]
		at org.glassfish.jersey.internal.Errors.process(Errors.java:244) ~[org.glassfish.jersey.core-jersey-common-2.34.jar:?]
		at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:288) ~[org.glassfish.jersey.core-jersey-common-2.34.jar:?]
		at org.glassfish.jersey.client.ClientRuntime$2.failure(ClientRuntime.java:178) ~[org.glassfish.jersey.core-jersey-client-2.34.jar:?]
		at org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector.lambda$apply$1(AsyncHttpConnector.java:223) ~[io.streamnative-pulsar-client-admin-original-2.10.3.6.jar:2.10.3.6]
		at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
		at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
		at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
		at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
		at org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector.lambda$retryOperation$4(AsyncHttpConnector.java:282) ~[io.streamnative-pulsar-client-admin-original-2.10.3.6.jar:2.10.3.6]
		at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
		at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
		at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
		at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
		at org.asynchttpclient.netty.NettyResponseFuture.abort(NettyResponseFuture.java:273) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.request.NettyRequestSender.abort(NettyRequestSender.java:473) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.request.NettyRequestSender$1.onFailure(NettyRequestSender.java:325) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.SimpleFutureListener.operationComplete(SimpleFutureListener.java:26) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:557) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:492) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.util.concurrent.DefaultPromise.addListener(DefaultPromise.java:185) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.util.concurrent.DefaultPromise.addListener(DefaultPromise.java:35) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at org.asynchttpclient.netty.request.NettyRequestSender.sendRequestWithNewChannel(NettyRequestSender.java:301) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.request.NettyRequestSender.sendRequestWithCertainForceConnect(NettyRequestSender.java:142) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.request.NettyRequestSender.sendRequest(NettyRequestSender.java:113) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.request.NettyRequestSender.sendNextRequest(NettyRequestSender.java:548) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.handler.intercept.Redirect30xInterceptor.exitAfterHandlingRedirect(Redirect30xInterceptor.java:160) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.handler.intercept.Interceptors.exitAfterIntercept(Interceptors.java:98) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.handler.HttpHandler.handleHttpResponse(HttpHandler.java:78) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.handler.HttpHandler.handleRead(HttpHandler.java:140) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at org.asynchttpclient.netty.handler.AsyncHttpClientHandler.channelRead(AsyncHttpClientHandler.java:78) ~[org.asynchttpclient-async-http-client-2.12.1.jar:?]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[io.netty-netty-codec-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[io.netty-netty-codec-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[io.netty-netty-codec-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[io.netty-netty-transport-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[io.netty-netty-common-4.1.87.Final.jar:4.1.87.Final]
		at java.lang.Thread.run(Thread.java:829) ~[?:?]
	Caused by: java.util.concurrent.CompletionException: org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector$RetryException: Could not complete the operation. Number of retries has been exhausted. Failed reason: pulsar-broker-2.pulsar-broker.pulsar.svc.cluster.local
		at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331) ~[?:?]
		at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346) ~[?:?]
		at java.util.concurrent.CompletableFuture$OrApply.tryFire(CompletableFuture.java:1503) ~[?:?]
		... 52 more
```

File: charts/pulsar/values.yaml (hunk 4 above)
Comment (ericsyh) [reply charts/pulsar/values.yaml:556]: Not very sure about this part, just keep consistent with https://github.com/apache/pulsar-helm-chart/blob/master/charts/pulsar/values.yaml#L349.
But the weird thing is even if not `securityContext` set on zk and bk, in my testing they are still okay to run.

File: charts/pulsar/values.yaml (hunk 3 above)
Comment (ericsyh) [reply charts/pulsar/values.yaml:533]: If no timeout set or with short time timeout, sometimes will get zk error like. Not sure about the RCA for this. 

```
12:07:55.432 [QuorumPeer[myid=1](plain=0.0.0.0:2181)(secure=disabled)] INFO  org.apache.zookeeper.server.quorum.FastLeaderElection - Notification time out: 51200
12:07:55.433 [QuorumConnectionThread-[myid=1]-1] WARN  org.apache.zookeeper.server.quorum.QuorumCnxManager - Cannot open channel to 3 at election address sn-pulsar-zookeeper-2.sn-pulsar-zookeeper.sn.svc.cluster.local:3888
java.net.UnknownHostException: sn-pulsar-zookeeper-2.sn-pulsar-zookeeper.sn.svc.cluster.local
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:229) ~[?:?]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:?]
	at java.net.Socket.connect(Socket.java:609) ~[?:?]
	at org.apache.zookeeper.server.quorum.QuorumCnxManager.initiateConnection(QuorumCnxManager.java:383) ~[io.streamnative-zookeeper-3.6.3.1.jar:3.6.3.1]
	at org.apache.zookeeper.server.quorum.QuorumCnxManager$QuorumConnectionReqThread.run(QuorumCnxManager.java:457) ~[io.streamnative-zookeeper-3.6.3.1.jar:3.6.3.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
```

File: charts/pulsar/templates/broker/broker-statefulset.yaml (hunk 1 above)
Comment (ericsyh) [reply charts/pulsar/templates/broker/broker-statefulset.yaml]: Removed in https://github.com/streamnative/charts/pull/985/commits/85d06b27346c1651db1940ae7b15c99a3b234d85 I think the error is from the file permission for `connectorsDirectory`.

**Comment by ericsyh** [comment]
@heesung-sn I saw the [PR](https://github.com/apache/pulsar-helm-chart/pull/180) on `pulsar-helm-chart` to change the zk serverCnxnFactory from `org.apache.zookeeper.server.NettyServerCnxnFactory` to `org.apache.zookeeper.server.NIOServerCnxnFactory`. Shall we follow this change? But in my testing seems no impact if keep using `org.apache.zookeeper.server.NettyServerCnxnFactory`

**Comment by heesung-sn** [comment]
Yes, I think we need to use the `NIOServerCnxnFactory` as it's a bug fix.

**Review by heesung-sn — APPROVED** [review]

**Review by maxsxu — APPROVED** [review]

**Pull request merged**